docs/docs/adding-tags-and-categories-to-blog-posts.md

```diff
@@ -54,230 +64,238 @@ query IndexQuery {
-    createPage({
-      path: `/tags/${tagName}`,
-      component: tagTemplate,
-      context: {
-        posts,
-        post,
-        tag: tagName,
-      },
-    });
-  });
-};
-```
+## Modify `gatsby-node.js` to render pages using that template
 
-Then in the `createPages` API function, query using `graphql` for your fields and use that to call the new `createTagPages` function.
+Now we've got a template. Great! I'll assume you followed the tutorial for [Adding Markdown Pages](/docs/adding-tags-and-categories-to-blog-posts/) and provide a sample `createPages` that generates post pages as well as tag pages. In the site's `gatsby-node.js` file, include `lodash` (`const _ = require('lodash')`) and then make sure your [`createPages`](/docs/node-apis/#createPages) looks something like this:
 
-```javascript
-exports.createPages = ({ graphql, boundActionCreators }) => {
+```js
+exports.createPages = ({ boundActionCreators, graphql }) => {
   const { createPage } = boundActionCreators;
-  // add the tags to the query
-  return new Promise((resolve, reject) => {
-    graphql(`
-      {
-        allMarkdownRemark {
-          edges {
-            node {
-              fields {
-                slug
-              }
-              frontmatter {
-                tags
-                title
-              }
+
+  const blogPostTemplate = path.resolve('src/templates/blog.js');
+  const tagTemplate = path.resolve('src/templates/tags.js');
+
+  return graphql(`
+    {
+      allMarkdownRemark(
+        sort: { order: DESC, fields: [frontmatter___date] }
+        limit: 2000
+      ) {
+        edges {
+          node {
+            frontmatter {
+              path
+              tags
             }
           }
         }
       }
-    `).then(result => {
-      console.log(result);
-      const posts = result.data.allMarkdownRemark.edges;
-
-      // call createTagPages with the result of posts
-      createTagPages(createPage, posts);
-
-      // this is the original code used to create the pages from markdown posts
-      result.data.allMarkdownRemark.edges.map(({ node }) => {
-        createPage({
-          path: node.fields.slug,
-          component: path.resolve(`./src/templates/blog-post.js`),
-          context: {
-            slug: node.fields.slug,
-          },
-        });
+    }
+  `).then(result => {
+    if (result.errors) {
+      return Promise.reject(result.errors);
+    }
+
+    const posts = result.data.allMarkdownRemark.edges;
+
+    // Create post detail pages
+    posts.forEach(({ node }) => {
+      createPage({
+        path: node.frontmatter.path,
+        component: blogPostTemplate,
+      });
+    });
+
+    // Tag pages:
+    let tags = [];
+    // Iterate through each post, putting all found tags into `tags`
+    _.each(posts, edge => {
+      if (_.get(edge, 'node.frontmatter.tags')) {
+        tags = tags.concat(edge.node.frontmatter.tags);
+      }
+    });
+    // Eliminate duplicate tags
+    tags = _.uniq(tags);
```

Comment:
I've been using a set for this lately, works quite well!

```js
const tags = posts.reduce((uniq, { node }) => {
  (node.frontmatter.tags || [])
    .forEach(uniq.add, uniq);
   return uniq;
}, new Set());

// tags.forEach as previously
```

Reply:
Set needs polyfilled still https://kangax.github.io/compat-table/es6/#test-Set

You can do this cool trick :-) `new Set([...node.frontmatter.tags])`

Reply:
Oh, this is Node code 😅 no polyfilling needed here